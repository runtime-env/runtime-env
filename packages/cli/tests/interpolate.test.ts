import fs from "fs";
import path from "path";
import { spawnSync } from "child_process";
import { tmpdir } from "tmp";

const envSchemaFilePath = path.resolve(__dirname, ".runtimeenvschema.json");
const requiredEnv = {
  FOO: "inline",
};
const requiredEnvInvalid = {};
const optionalEnv = {
  FOO: "inline",
  BAR: '{"BAZ":"baz"}',
};
const optionalEnvInvalid = {
  FOO: "inline",
  BAR: "invalid",
};

// write integration tests here for interpolate command
describe("integration - interpolate", () => {
  test("globalVariableName", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--global-variable-name",
        "env",
        "--",
        "123 <%= env.FOO %> 456",
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...requiredEnv,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
  });

  test("envFileSchemaPath", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--env-schema-file-path",
        envSchemaFilePath,
        "--",
        "123 <%= runtimeEnv.FOO %> 456",
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...requiredEnv,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
  });

  test("inputFilePath", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--input-file-path",
        path.resolve(__dirname, "input-file.html"),
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...requiredEnv,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
  });

  test("inputFilePath - invalid", () => {
    const result = spawnSync(
      "node",
      ["../bin/runtime-env.js", "interpolate", "--input-file-path", "invalid"],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...requiredEnv,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(1);
  });

  test("outputFilePath", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--output-file-path",
        path.resolve(tmpdir, "runtime-env-interpolate-output-file-path.html"),
        "--",
        "123 <%= runtimeEnv.FOO %> 456",
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...requiredEnv,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
    expect(
      fs.readFileSync(
        path.resolve(tmpdir, "runtime-env-interpolate-output-file-path.html"),
        "utf8",
      ),
    ).toMatchSnapshot();
  });

  test("requiredEnv", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--",
        "123 <%= runtimeEnv.FOO %> 456\\n 123 <%= runtimeEnv.BAR.BAZ %> 456",
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...requiredEnv,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
  });

  test("requiredEnv - invalid", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--",
        "123 <%= runtimeEnv.FOO %> 456\\n 123 <%= runtimeEnv.BAR.BAZ %> 456",
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...requiredEnvInvalid,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(1);
  });

  test("optionalEnv", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--",
        "123 <%= runtimeEnv.FOO %> 456\\n 123 <%= runtimeEnv.BAR.BAZ %> 456",
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...optionalEnv,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(0);
    expect(result.output).toMatchSnapshot();
  });

  test("optionalEnv - invalid", () => {
    const result = spawnSync(
      "node",
      [
        "../bin/runtime-env.js",
        "interpolate",
        "--",
        "123 <%= runtimeEnv.FOO %> 456\\n 123 <%= runtimeEnv.BAR.BAZ %> 456",
      ],
      {
        encoding: "utf8",
        stdio: "pipe",
        env: {
          ...process.env,
          ...optionalEnvInvalid,
        },
        cwd: __dirname,
      },
    );
    expect(result.status).toBe(1);
  });
});
