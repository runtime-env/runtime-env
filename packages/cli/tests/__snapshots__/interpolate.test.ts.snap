// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration - interpolate envFileSchemaPath 1`] = `
[
  null,
  "123 inline 456
",
  "",
]
`;

exports[`integration - interpolate globalVariableName 1`] = `
[
  null,
  "123 inline 456
",
  "",
]
`;

exports[`integration - interpolate inputFilePath - invalid 1`] = `
[
  null,
  "",
  "[@runtime-env/cli]
  input file not found: no such file, open 'invalid'
",
]
`;

exports[`integration - interpolate inputFilePath 1`] = `
[
  null,
  "123 inline 456

",
  "",
]
`;

exports[`integration - interpolate optionalEnv - invalid 1`] = `
[
  null,
  "",
  "[@runtime-env/cli]
  env is invalid: [{"instancePath":"/BAR","schemaPath":"#/properties/BAR/type","keyword":"type","params":{"type":"object"},"message":"must be object"}]
",
]
`;

exports[`integration - interpolate optionalEnv 1`] = `
[
  null,
  "123 inline 456\\n 123 baz 456
",
  "",
]
`;

exports[`integration - interpolate outputFilePath 1`] = `
[
  null,
  "",
  "",
]
`;

exports[`integration - interpolate outputFilePath 2`] = `"123 inline 456"`;

exports[`integration - interpolate requiredEnv - invalid 1`] = `
[
  null,
  "",
  "[@runtime-env/cli]
  env is invalid: [{"instancePath":"","schemaPath":"#/required","keyword":"required","params":{"missingProperty":"FOO"},"message":"must have required property 'FOO'"}]
",
]
`;

exports[`integration - interpolate requiredEnv 1`] = `
[
  null,
  "123 inline 456\\n 123  456
",
  "",
]
`;
